core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -128,21 +137,90 @@
   }
 
   /**
+   * An interface extending {@link FieldIndex}, represents a unique index. A Unique index is an
+   * index where each index value only maps to one object.
+   *
+   * @param <T> type of objects in this {@link IndexedSet}
+   */
+  public interface UniqueFieldIndex<T> extends FieldIndex<T> {
+  }
+
+  /**
+   * An interface extending {@link FieldIndex}, represents a non-unique index. A non-unique index is
+   * an index where a index value can map to one or more objects.
+   *
+   * @param <T> type of objects in this {@link IndexedSet}
+   */
+  public interface NonUniqueFieldIndex<T> extends FieldIndex<T> {
+  }
+
+  /**
    * Constructs a new {@link IndexedSet} instance with at least one field as the index.
    *
    * @param field at least one field is needed to index the set of objects
    * @param otherFields other fields to index the set
    */
   @SafeVarargs
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
-        new HashMap<>(otherFields.length + 1);
-    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+    // count the numbers of two index types
+    int uniqueIndexLength = 0;
+    int nonUniqueIndexLength = 0;
+
+    if (field instanceof UniqueFieldIndex) {
+      uniqueIndexLength = 1;
+    } else {
+      nonUniqueIndexLength = 1;
+    }
+
+    for (FieldIndex<T> fieldIndex : otherFields) {
+      if (fieldIndex instanceof UniqueFieldIndex) {
+        uniqueIndexLength++;
+      } else {
+        nonUniqueIndexLength++;
+      }
+    }
+
+    // initialization
+    Map<NonUniqueFieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMapNonUnique =
+        null;
+    Map<UniqueFieldIndex<T>, ConcurrentHashMap<Object, T>> indexMapUnique = null;
+    if (uniqueIndexLength > 0) {
+      indexMapUnique = new HashMap<>(uniqueIndexLength);
+    }
+    if (nonUniqueIndexLength > 0) {
+      indexMapNonUnique = new HashMap<>(nonUniqueIndexLength);
+    }
+
+    if (field instanceof UniqueFieldIndex) {
+      indexMapUnique.put((UniqueFieldIndex) field, new ConcurrentHashMap<Object, T>(8, 0.95f, 8));
+    } else {
+      indexMapNonUnique.put((NonUniqueFieldIndex<T>) field,
+          new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+    }
+
     for (FieldIndex<T> fieldIndex : otherFields) {
```

Comment:
@aaudiber Maybe we cannot. We should initialize the two indexMaps before we can run the second for loop.